### ISSUE TITLE
Implement a set of built-in methods to be used on strings/integers/....

### ISSUE THREAD

**Issue by namannimmo10** [lcompilers/lpython#281]
Since we now have the [`Attribute` visitor](https://github.com/lcompilers/lpython/blob/main/src/lpython/semantics/python_ast_to_asr.cpp#L1844) in AST->ASR, we should start implementing the set of built-in methods to be used on strings, integers, lists, dictionaries, etc.

Here's a list of those operating on strings: https://www.w3schools.com/python/python_ref_string.asp
And one can navigate on the left for other types.

Let's start with something like `bit_length` for integers:
```python
>>> num = 9
>>> num.bit_length()
4
```

**Comment by certik** [comment]
Yes. Implement them as `_lpython_bit_length(i: i32)`, make the AST to ASR insert these in. Those that can be implemented in Python should be. Otherwise just do the declaration and the backend (ASR->LLVM) has to implement them.

**Comment by namannimmo10** [comment]
We can try this:
```python
>>> def _lpython_bit_length(i: int) -> int:
...   x: str
...   x = bin(i)
...   if i == 0: return 0
...   elif i < 0: x = x[3:]
...   else: x = x[2:]
...
...   return len(x)
...
>>> all(_lpython_bit_length(i) == i.bit_length() for i in range(-1000, 1000))
True
>>>
```

**Comment by Smit-create** [comment]
Just to make the design cleaner and not just put everything into `Attribute`, I suggest creating a map something similar to intrinsic procedures, where we can directly look for the methods (for example, maybe we can map `list.append` to a string `list%append` and use this converted string to find the appropriate function.)

**Comment by certik** [comment]
We should try to systematically manage these methods, and at the same time keep the complexity low at the C++ level. Having some kind of a map like we do for `comptime_eval` is fine I think.

**Comment by Shaikh-Ubaid** [comment]
Please, could someone possibly share if these built in methods need to be implemented as compile time functions (like [here](https://github.com/lcompilers/lpython/blob/64e0c33ca22e4cf0bf21c56fd78286ee1721e51d/src/lpython/semantics/python_comptime_eval.h#L100) ) or runtime functions (like [here](https://github.com/lcompilers/lpython/blob/64e0c33ca22e4cf0bf21c56fd78286ee1721e51d/src/runtime/lpython_builtin.py#L52)) or both?

**Comment by certik** [comment]
@Shaikh-Ubaid every intrinsic function has both compile time and runtime implementation. Sorry for my late reply, I was just going through my notifications and noticed your question.

**Comment by Shaikh-Ubaid** [comment]
Got it. Thank you for the clarification!

**Comment by certik** [comment]
Add the function `_lpython_bit_length` into `src/runtime/lpython_builtin.py`, then in AST->ASR, when you encounter `num.bit_length()`, simply call this function.

**Comment by namannimmo10** [comment]
@Abdelrahman-Kh-Fouad -- are you working on this?

**Comment by Abdelrahman-Kh-Fouad** [comment]
> @Abdelrahman-Kh-Fouad -- are you working on this?

yes i will, after `str()` casting.

**Comment by navneet1v** [comment]
Do we still have any methods left to implement? I would like to implement.

**Comment by certik** [comment]
@navneet1v thanks for your interest! 

If I am not mistaken, all the methods on integer that we implemented are here: https://github.com/lcompilers/lpython/blob/b3534ff898aa3e948f10189633d83138815b48a8/src/lpython/semantics/python_ast_to_asr.cpp#L4728, and it seems only `bit_length` is currently implemented.

So if you are interested in implementing any of the other ones, that would be awesome. Let me know if you have any questions or need help getting started. I am happy to meet to give you an overview.

**Comment by zeproffesor** [comment]
Hello, I am trying to implement the to_bytes method for integers. As mentioned here [https://docs.python.org/3/library/stdtypes.html](url) the method returns an array of bytes, which is a type, i.e. `print((5).to_bytes(1,'big'))` --> `<class 'bytes'>` 

My question is that do we want to implement:
1. A new type, i.e. implement the class 'bytes'
2. Just print the string corresponding to the byte array, i.e. `print((1023).to_bytes(2,'big))` --> `b'\x03\xff'`
3. Instead of a byte array make a list that corresponds to the byte array, i.e. `print((1023).to_bytes(2,'big))` --> `[3,255]`
4. Something else 

Also I will appreciate it if you would like to add any other tips that can help me out! Thanks!

**Comment by certik** [comment]
Thanks @zeproffesor, great question. I opened up a dedicated issue (https://github.com/lcompilers/lpython/issues/1187) to discuss the bytes type. 

Until then just return a string. Once we decided how to handle the bytes type, we can upgrade it.

**Comment by zeproffesor** [comment]
Okay! Thanks for the clarification, I'll begin working on the string implementation then.

**Comment by certik** [comment]
Thanks! Feel free to open a Draft PR and post what you have and we can help if you get stuck.

**Comment by zeproffesor** [comment]
@certik Created a draft pull request. Please review! Thanks!

**Comment by rsaba5612** [comment]
bit_length() is a built-in method in Python that returns the number of bits required to represent an integer in binary, excluding the sign and leading zeros.

For example, the binary representation of 9 is 0b1001, which has a length of 4 (excluding the leading 0's). Therefore, num.bit_length() returns 4.

Here's an example:

num = 9
bit_length = num.bit_length()
print(bit_length)  # output: 4
You can use bit_length() method on any integer in Python.

**Comment by OmarMesqq** [comment]
Hi! Is anyone still working on this? I am quite new to compilers but I plan on helping you guys :) 
Project looks promising!

I am working on a PR for this issue

**Comment by certik** [comment]
@OmarMesqq perfect! Go ahead and work on it. A lot of these are implemented, but I think some are still missing. If you need any help, please ask!
